python/orca/src/bigdl/orca/learn/pytorch/callbacks/model_checkpoint.py

```diff
@@ -0,0 +1,126 @@
+        self.save_weights_only = save_weights_only
+        self.last_ckpt_path = ""
+
+    def on_batch_begin(self, batch):
+        """
+        Called at the beginning of a training batch in `fit` methods.
+        Subclasses should override for any actions to run.
+        @param batch: Integer, index of batch within the current epoch.
+        """
+        pass
+
+    def on_batch_end(self, batch):
+        """
+        Called at the end of a training batch in `fit` methods.
+        Subclasses should override for any actions to run.
+        @param batch: Integer, index of batch within the current epoch.
+        """
+        pass
+
+    def on_epoch_begin(self, epoch):
+        """
+        Called at the start of an epoch.
+        Subclasses should override for any actions to run. This function should only
+        be called during TRAIN mode.
+        @param epoch: Integer, index of epoch.
+        @param logs: Dict. Currently, saved stats in last epoch has been passed to this argument
+        for this method but may change in the future.
+        """
+        pass
+
+    def on_epoch_end(self, epoch):
+        """
+        Called at the end of an epoch.
+        Subclasses should override for any actions to run. This function should only
+        be called during TRAIN mode.
+        @param epoch:  Integer, index of epoch.
+        """
+        stats = {"epoch": self.trainer.epochs}
+        last_ckpt_path = self._format_checkpoint_name(self.filepath, stats)
+        self.trainer.save_checkpoint(last_ckpt_path, self.save_weights_only)
+
+    def on_train_begin(self):
+        """
+        Called at the beginning of training.
+        Subclasses should override for any actions to run.
+        @param logs: Dict. Currently, no data is passed to this argument for this method
+          but that may change in the future.
+        """
+        # todo: support resume training
+        from bigdl.orca.learn.pytorch.utils import get_filesystem
+        dirname = os.path.dirname(self.filepath)
+        fs = get_filesystem(dirname)
+        if fs.exists(dirname):
+            files = [os.path.basename(f["name"]) for f in fs.listdir(dirname)]
+            files = [x for x in files if "ckpt" in x]
+            if len(files) == 0:
+                return None
+            raise ValueError(f"Find non-empty dirname with filepath of {self.filepath}.")
+        else:
+            fs.mkdirs(dirname)
+
+    def on_train_end(self):
+        """
+        Called at the end of training.
+        Subclasses should override for any actions to run.
+        """
+        stats = {"epoch": self.trainer.epochs}
+        last_ckpt_path = self._format_checkpoint_name(self.filepath, stats, last=True)
+        previous, self.last_ckpt_path = self.last_ckpt_path, last_ckpt_path
+        self.trainer.save_checkpoint(last_ckpt_path, self.save_weights_only)
+        if previous and previous != last_ckpt_path:
+            self.trainer.remove_checkpoint(previous)
```

Comment:
maybe we do not need to remove previous ones?

Reply:
It would be helpful if you could elaborate on your concerns.

Reply:
I think it is more common to save the checkpoint in every epoch. For example, if the training become unstable after certain epochs, users might need to return back a few epochs to recover. Or maybe some epoch in the middle produce the best model.

Reply:
Yeah, checkpoints for every epoch have been saved in `on_epoch_end` and will not be removed. And I have also saved the last checkpoint after all train ends. The removal is for the previous **last** checkpoint (if exists).

Reply:
I see